---
author: "Peter Diakumis (UMCCR)"
date: "`r format(Sys.time(), '%a %Y-%b-%d')`"
output:
  html_document:
    keep_md: yes
    theme: readable
    toc: false
    toc_float: false
    code_folding: hide
  rmdformats::material:
    highlight: kate
params:
  batchname: "x"
  cval: 123
  results_dir: "x"
  title: "Facets Summary"
title: "`r paste(params$batchname, params$title)`"
---

```{r load-pkgs, message=FALSE}
require(dplyr)
require(readr)
require(devtools)
require(DT)
```

```{r set-params-interactive, eval=FALSE, echo=FALSE}
params <- list(batchname = "HCC2218", cval = "150", results_dir = "/data/cephfs/punim0010/projects/Diakumis_HCC2218/woof/final/structural/facets/results")
```

```{r funcs}
cnv_tab <- function(cnv) {
  bigmark_col <- function(x) {
    format(x, big.mark = ",", scientific = FALSE)
  }
  
  DT::datatable(cnv %>%
                  dplyr::mutate_if(is.numeric, bigmark_col),  
                options = list( 
                  columnDefs = list(list(className = 'dt-right', targets = c(3, 4)))),
                filter = "top",
                rownames = FALSE)
}
```


## Introduction

FACETS (https://github.com/mskcc/facets) is an R package that implements the
methods described in [Shen and Seshan, 2016](https://www.ncbi.nlm.nih.gov/pubmed/27270079).
From the vignette:

>  This package is for analyzing allele-specific DNA copy number and clonal heterogeneity
from high-throughput sequencing including whole genome, whole exome, and some
targeted cancer gene panels. The method implements a bivariate genome segmentation,
followed by allele-specific copy number calls. Tumor purity, ploidy, and cellular
fractions are estimated and reported from the output.

## Parameters

* Batch Name: `r params$batchname`
* cval: `r params$cval`
* Results directory: `r params$results_dir`


## Results

### Purity and Ploidy Estimation


```{r read-data}
results_dir <- file.path(params$results_dir, params$batchname)
prefix <- paste0(params$batchname, "_cval_", params$cval)
purply <- readr::read_tsv(file.path(results_dir,  paste0(prefix, "_purply.tsv")), col_types = "dddd")
cnv <- readr::read_tsv(file.path(results_dir, paste0(prefix, "_segs.tsv")), col_types = "ciiiddiddiidii") %>% 
  dplyr::select(seg, chrom, start, end, tcn.em, lcn.em, dplyr::everything())
```

```{r output-purity-ploidy}
knitr::kable(purply)
```


### CNV Plot

```{r cnv-plot, out.width = "100%"}
plot_type <- "cnv"
fname <- file.path(results_dir, paste0(prefix, "_", plot_type, ".png"))
knitr::include_graphics(fname)
```

### Spider Plot

```{r spider-plot, out.width = "80%"}
plot_type <- "spider"
fname <- file.path(results_dir, paste0(prefix, "_", plot_type, ".png"))
knitr::include_graphics(fname)
```

### CNV Table

```{r cnv-table}
cnv_tab(cnv)
```

## Session Info

```{r session-info}
devtools::session_info()
```
