---
author: "Peter Diakumis (UMCCR)"
date: "`r format(Sys.time(), '%a %Y-%b-%d')`"
output:
  html_document:
    keep_md: yes
    theme: readable
    toc: true
    toc_float: true
    code_folding: hide
  rmdformats::material:
    highlight: kate
params:
  tumor_name: "x"
  results_dir: "x"
  title: "Purple Summary"
title: "`r paste0(params$tumor_name, '\n\n', params$title)`"
---

```{r load-pkgs, message=FALSE}
require(dplyr)
require(readr)
require(glue)
require(devtools)
require(DT)
require(rock, lib.loc = "/Users/pdiakumis/rock/master")
```

```{r render_interactively, eval=FALSE, echo=FALSE}
render_me <- function() {
  rmarkdown::render("purple_report.Rmd",
                    params = list(tumor_name = "PRJ180467_P045_FFPE_tumour", 
                                  results_dir = "/Users/pdiakumis/Desktop/projects/umccr/tothill_projects/data/collibri/purple/P045")
                    )
}

render_me()
```

```{r funcs}
cnv_tab <- function(cnv) {
  
  DT::datatable(cnv, filter = "top", rownames = FALSE) %>% 
    DT::formatCurrency(~ start + end + seg_len, currency = "", interval = 3, mark = ",", digits = 0) %>% 
    DT::formatRound('tot_cn')
}
```




[PURPLE](https://github.com/hartwigmedical/hmftools/tree/master/purity-ploidy-estimator)
is a tool for estimating tumor purity and ploidy. It uses the read depth and 
tumor beta allele frequency (BAF) to estimate purity and generate a copy number profile.

### Parameters

* Tumor Name: ``r params$tumor_name``
* Results directory: ``r params$results_dir``

## Circos

### Plot 1: Somatic SNVs/Indels, Total/Minor CN, SVs

* __Track1__: Chromosomes. Darker shaded areas: gaps in reference genome 
  (centromeres, heterochromatin & missing short arms)
* __Track2__: Somatic variants (incl. exon, intron and intergenic regions). 
    * outer ring: SNP allele frequencies, corrected for tumor purity and scaled from 0 to 100%.
      Each dot represents a single somatic variant, coloured according to the
      type of base change (e.g. C>T/G>A in red).
    * inner ring: short insertion (yellow) and deletion (red) locations. 
* __Track3__: Observed copy number changes adjusted for tumor purity,
  including focal and chromosomal somatic events. 
  <span style="color:red">Red</span> = Loss; <span style="color:#32CD32">Green</span> = Gain.
  Scaled from 0 (complete loss) to 6 (high level gains). 
  If > 6, shown as 6 with a green dot on the outermost green gridline. 
* __Track4__: Observed 'minor allele copy numbers'. Range from 0 to 3. 
  Expected normal minor allele copy number is 1, and anything below 1 is shown 
  as a loss (<span style="color:#EE7600">Orange</span>), representing an LOH event. 
  Minor allele copy numbers above 1 (<span style="color:#7EC0EE">Blue</span>) indicate gains 
  of both A and B alleles. 
* __Track5__ (Inner circle): Observed structural variants within or between the chromosomes.
    * <span style="color:#7EC0EE">Blue</span> = Translocations
    * <span style="color:red">Red</span> = Deletions 
    * <span style="color:#e6e600">Yellow</span> = Insertions 
    * <span style="color:#32CD32">Green</span> = Tandem duplications 
    * <span style="color:#000000">Black</span> = Inversions

```{r circos-default1, out.width="80%"}
plot_path <- file.path(params$results_dir, "plot") 
purple_circos_dir <- file.path(params$results_dir, "circos")
knitr::include_graphics(file.path(plot_path, paste0(params$tumor_name, ".circos.png")))
```

### Plot 2: Tumor/Normal Allele Ratios, BAF

* __Track1__: Chromosomes. Darker shaded areas: gaps in reference genome 
  (centromeres, heterochromatin & missing short arms)
* __Track2__: <span style="color:blue">Tumor</span> and <span style="color:#32CD32">Normal</span> Ratios
* __Track3__: <span style="#EE7600">Beta Allele Frequency</span>


```{r circos-default2, out.width="80%"}
knitr::include_graphics(file.path(plot_path, paste0(params$tumor_name, ".input.png")))
```

### Plot 3: Total/Minor CN, SVs, BAF

```{r circos-baf-prep, eval=TRUE}
rock::plot_circos_baf(purple_dir = purple_circos_dir, out_dir = file.path(purple_circos_dir, "baf_circos"), tumor_name = params$tumor_name)
```

```{r circos-baf-plot}
knitr::include_graphics(file.path(purple_circos_dir, "baf_circos", paste0(params$tumor_name, "_circos_baf.png")))
```

## Purity and Ploidy Summary

```{r purply}
fn <- file.path(params$results_dir, paste0(params$tumor_name, ".purple.purity"))
x <- readr::read_tsv(fn, col_types = cols(.default = col_double(), 
                                          Gender = col_character(), 
                                          Status = col_character())) %>%
  dplyr::mutate_if(is.numeric, round, 2) %>%
  unlist()

tribble(~key, ~value,
        'Sex', x["Gender"],
        'Purity', glue('{x["#Purity"]} ({x["MinPurity"]}-{x["MaxPurity"]})'),
        'Ploidy', glue('{x["Ploidy"]} ({x["MinPloidy"]}-{x["MaxPloidy"]})'),
        'Diploidy Proportion', glue('{x["DiploidProportion"]} ({x["MinDiploidProportion"]}-{x["MaxDiploidProportion"]})'),
        'Polyclonal Proportion', x["PolyclonalProportion"],
        'Purple Version', x["Version"]) %>% 
  knitr::kable()
```

## AZ300 Gene Somatic CNV Calls

```{r gene-tab}
fn_gene <- file.path(params$results_dir, paste0(params$tumor_name, ".purple.gene.cnv"))
df <- readr::read_tsv(fn_gene, col_types = cols(.default = col_character()), skip = 1, col_names = FALSE) %>% 
  dplyr::select(chrom = X1, start = X2, end = X3, gene = X4, min_cn = X5, max_cn = X6) %>% 
  dplyr::mutate(start = as.integer(start), end = as.integer(end),
                min_cn = as.double(min_cn), max_cn = as.double(max_cn))

az300 <- readr::read_tsv(system.file("extdata", "az300.bed", package = "pebbles"), 
                         col_names = c("chrom", "start", "end", "gene"),
                         col_types = "ciic")

df %>% 
  filter(gene %in% az300$gene) %>% 
  DT::datatable(filter = "top", rownames = FALSE) %>% 
  DT::formatCurrency(~ start + end, currency = "", interval = 3, mark = ",", digits = 0) %>% 
  DT::formatRound(~ min_cn + max_cn)
```


## Global Somatic and Germline CNV Summary

```{r germ-som-prep}
fn_som <- file.path(params$results_dir, paste0(params$tumor_name, ".purple.cnv"))
fn_germ <- file.path(params$results_dir, paste0(params$tumor_name, ".purple.germline.cnv"))
cnv_som <- rock::read_cnv(fn_som)
cnv_germ <- readr::read_tsv(fn_germ, col_types = "ciiddddccc") %>%
  dplyr::select(chrom = `#chromosome`, start, end, tot_cn = copyNumber) %>% 
  dplyr::filter(.data$chrom != "MT")

cnv_germ <- structure(list(cnv = cnv_germ), class = "cnv")

cnv_all <- dplyr::bind_rows(list(somatic = cnv_som$cnv, germline = cnv_germ$cnv), .id = "type")
cnv_all %>% 
  mutate(sv_len = end - start) %>% 
  group_by(type) %>% 
  summarise(n = n(),
            mean_len = round(mean(sv_len), 2),
            med_len = round(median(sv_len), 2),
            min_cn = round(min(tot_cn), 2),
            max_cn = round(max(tot_cn), 2)) %>% 
  knitr::kable(format.args = list(big.mark = ","))
```

### Somatic Calls

```{r som-tab}
cnv_tab(cnv_som$cnv %>% mutate(seg_len = end - start))
```

## Other Stuff

```{r purple-other-plots, out.width="80%"}
knitr::include_graphics(file.path(plot_path, paste0(params$tumor_name, ".copyNumber.png")))
knitr::include_graphics(file.path(plot_path, paste0(params$tumor_name, ".minor_allele.png")))
knitr::include_graphics(file.path(plot_path, paste0(params$tumor_name, ".variant.png")))
```
