---
title: "Testing rr vcf functions"
author: "Peter Diakumis"
date: "`r format(Sys.time(), '%a %Y-%b-%d')`"
output: 
  html_document: 
    keep_md: yes
---

## Analysis 1

```{r load-pkgs}
require(OmicCircos)
require(dplyr)
require(readr)
```

* Which columns do we need to extract from the VCF?
    * CHROM
    * ID
    * FILTER
    * INFO/BPI_START
    * INFO/BPI_END
    * INFO/MATEID
    * INFO/SVTYPE
    
* How can we extract these?
    * `bcftools query -f "%CHROM\t%INFO/BPI_START\t%INFO/BPI_END\t%ID\t%INFO/MATEID\t%INFO/SVTYPE\t%FILTER\n" $vcf`
    
* Then:
    
```{r}
read_svs <- function(fname) {
  
  DF <- 
    readr::read_tsv(
      file = fname,
      na = c(".", "", "NA"),
      col_names = c("chrom", "bpi_start", "bpi_end", "id", "mateid", "svtype", "filter"), 
      col_types = "ciicccc") %>% 
    dplyr::mutate(
      rowid = 1:nrow(.), # used for sorting downstream
      chrom1 = chrom,
      start1 = bpi_start,
      end1 = bpi_start,
      END = bpi_end) %>%
    dplyr::select(rowid, chrom1, start1, end1, END, filter, id, mateid, svtype)
  
  # BNDs
  # just keep first BND mates (i.e. discard duplicated information)
  # see <https://github.com/Illumina/manta/blob/master/docs/developerGuide/ID.md>
  df_bnd <- DF %>%
    dplyr::filter(svtype == "BND") %>%
    dplyr::bind_cols(., .[match(.$id, .$mateid), c("chrom1", "start1")]) %>%
    dplyr::rename(chrom2 = chrom11, start2 = start11) %>%
    dplyr::mutate(end2 = start2,
                  bndid = substring(id, nchar(id))) %>% 
    dplyr::filter(bndid == "1")
  
  # Other
  df_other <- DF %>%
    dplyr::filter(svtype != "BND") %>%
    dplyr::mutate(chrom2 = chrom1, start2 = END, end2 = END)
  
  # All together now
  svs <- df_other %>%
    dplyr::bind_rows(df_bnd) %>%
    dplyr::select(rowid, chrom1:end1, chrom2:end2, svtype, id, filter) %>%
    dplyr::arrange(rowid)
  
  return(svs)
}
```
    
Then let's plot:

```{r}
#---- Chromosome data ----#
data("UCSC.hg19.chr", package = 'OmicCircos')
ucsc_chr <- UCSC.hg19.chr %>%
  mutate_if(is.factor, as.character) %>%
  mutate(chrom = sub("chr", "", chrom))

rm(UCSC.hg19.chr)

seg_name <- unique(ucsc_chr$chrom)
seg_num <- length(seg_name)

# Prepare angles + colors
db <- OmicCircos::segAnglePo(seg.dat = ucsc_chr, seg = seg_name)
colors <- rainbow(seg_num, alpha = 0.5)

#---- Manta ----#
mantaf1 <- "~/Desktop/tmp/E019-manta_svs.tsv"
svs <- read_svs(mantaf1)

# Now we can filter, select etc.
svs <- svs %>% 
  dplyr::filter(filter == "PASS") %>% 
  dplyr::mutate(varnum = paste0("var", rowid),
                varnum2 = varnum) %>% 
  dplyr::select(chrom1, start1, varnum,
                chrom2, start2, varnum2, svtype) %>% 
  dplyr::mutate(col = dplyr::case_when(
      svtype == "DEL" ~ "red",
      svtype == "DUP" ~ "green",
      svtype == "INS" ~ "purple",
      svtype == "INV" ~ "orange",
      TRUE    ~ "pink"))

svs_bnd <- svs %>% filter(svtype == "BND") %>% as.data.frame()
svs_other <- svs %>% filter(svtype != "BND") %>% as.data.frame()

#---- Facets ----#
facets_fname <- "~/Desktop/projects/umccr/A5/batch1/E019/E019_cval_150_fit.rds"
facets_cnv <- readr::read_rds(facets_fname)$cncf %>% 
  dplyr::select(chrom, start, end, tcn.em) %>%
  dplyr::rename(chr = chrom,
                CN = tcn.em) %>% 
  dplyr::mutate(
    col = dplyr::case_when(
      CN == 0 ~ "red",
      CN == 1 ~ "red",
      CN == 2 ~ "black",
      CN >= 3 ~ "green",
      TRUE    ~ "orange"))

#---- Circos Plot ----

png("~/Desktop/tmp/circos1.png", width = 1000, height = 1000)
par(mar = c(2, 2, 2, 2))
plot(c(1, 850), c(1, 850), type = "n", axes = FALSE, xlab = "", ylab = "", main = "")

circos(R = 400, cir = db, type = "chr", col = colors, print.chr.lab = TRUE, W = 4, scale = TRUE, cex = 3)
circos(R = 260, cir = db, type = "arc", W = 120, mapping = facets_cnv, col.v = 4, B = TRUE, cutoff = 2, lwd = 4, col = facets_cnv$col, scale = TRUE, cex = 10)

circos(R = 260, cir = db, type = "link", W = 40, mapping = svs_bnd, lwd = 2, col = "grey")
circos(R = 260, cir = db, type = "link2", W = 20, mapping = svs_other, lwd = 1, col = svs_other$col)
dev.off()
```


    



* Ideogram object

```{r gr-ideo}
# Generates GRanges hg19 or b37 Ideogram object
gr_ideo <- function(DF) {
  # check if you have hg19 or b37
  chr_names_hg19 <- paste0("chr", c(1:22, "X", "Y", "M"))
  chr_names_b37 <- c(1:22, "X", "Y", "MT")

  stopifnot(all(c("chrom1", "chrom2") %in% colnames(DF)))
  df_chr <- unique(unlist(DF[c("chrom1", "chrom2")], use.names = FALSE))
  if (all(df_chr %in% chr_names_b37)) {
    chr_names = chr_names_b37
  } else if (all(df_chr %in% chr_names_hg19)) {
    chr_names = chr_names_hg19
  } else {
    stop("ERROR! Can't determine if you have b37 or hg19 chromosomes. Check manually!")
  }

  chr_lengths <- c(249250621L, 243199373L, 198022430L, 191154276L,
                   180915260L, 171115067L, 159138663L, 146364022L,
                   141213431L, 135534747L, 135006516L, 133851895L,
                   115169878L, 107349540L, 102531392L,  90354753L,
                   81195210L,  78077248L,   59128983L,  63025520L,
                   48129895L,  51304566L,  155270560L,  59373566L, 16569L)
  names(chr_lengths) <- chr_names
  ideo <- GenomicRanges::GRanges(seqnames = chr_names,
                  ranges = IRanges::IRanges(start = 1L, width = chr_lengths),
                  seqlengths = chr_lengths)
  return(ideo)
}
```

* GRanges from input VCF

```{r df2gr}
# Generates GRanges object for input Df
df2gr <- function(DF, ideo) {
  sv_gr <- GenomicRanges::GRanges(
    seqnames = dplyr::pull(DF, chrom1),
    ranges = IRanges::IRanges(
      dplyr::pull(DF, start1),
      dplyr::pull(DF, end1)),
    SVTYPE = dplyr::pull(DF, SVTYPE),
    to_gr = GenomicRanges::GRanges(dplyr::pull(DF, chrom2),
                                   IRanges::IRanges(dplyr::pull(DF, start2),
                                                    dplyr::pull(DF, end2))))
                                                    
  
  GenomeInfoDb::seqlevels(sv_gr) <- GenomeInfoDb::seqlevels(ideo)
  GenomeInfoDb::seqlengths(sv_gr) <- GenomeInfoDb::seqlengths(ideo)
  GenomeInfoDb::seqlevels(sv_gr$to_gr) <-  GenomeInfoDb::seqlevels(ideo)
  GenomeInfoDb::seqlengths(sv_gr$to_gr) <- GenomeInfoDb::seqlengths(ideo)

  return(sv_gr)
}
```

* Circos from input VCF and Ideogram

```{r}
# Generates circos plot for Manta input
manta_plot <- function(gr_sv, gr_ideo, sample_nm) {

  cols <- scales::hue_pal()(length(unique(values(gr_sv)$SVTYPE)))

  p1 <- ggbio() +
    circle(gr_ideo, geom = "ideo", colour = "gray70", fill = "gray70", trackWidth = 0.1, radius = 3) +
    circle(gr_ideo, geom = "text", aes(label = seqnames), trackWidth = 1, radius = 3) +
    circle(gr_sv, geom = "link", linked.to = "to_gr", aes(color = SVTYPE), radius = 2.5, trackWidth = 1) +
    scale_color_manual(values = cols) +
    labs(title = glue("{sample_nm} Manta {length(gr_sv)} filtered calls"),
         subtitle = "Coloured by SVTYPE")

  return(p1)
}
```

* Results:

```{r}
ideo <- gr_ideo(svs)
gr_sv <- df2gr(svs, ideo)
manta_plot(gr_sv, ideo, samplename)
```

