# Validate NGS data

Step 1
------

* Use `woofr::validation_outputs` to generate a TSV file with the following columns:
  - `col1` - type of file
    1. VCF
    2. BAM
    3. FASTQ
    4. Other
  - `col2` - path to file

Step 2
------

* Now we can iterate through each row of the TSV according to its type:
  - `VCF`:
    - `head` 
    - [vcf-validator](https://github.com/EBIvariation/vcf-validator)
  - `BAM`: 
    - [samtools quickcheck](http://www.htslib.org/doc/samtools.html)
    - GATK ValidateSamFile [link1](https://software.broadinstitute.org/gatk/documentation/article.php?id=7571),
  [link2](http://broadinstitute.github.io/picard/command-line-overview.html#ValidateSamFile)
    - UMich [BamUtil validate](https://genome.sph.umich.edu/wiki/BamUtil:_validate)
  - `FASTQ`: 
    - [fqtools validate](https://github.com/alastair-droop/fqtools)

Output
------

```
woof-validation1/final
    |--- vcf/
         |--- tool_1/
              |--- results_fileA.txt
              |--- results_fileB.txt
         |--- tool_2/
              |--- results_fileA.txt
              |--- results_fileB.txt
    |--- bam/
         |--- tool_1/
              |--- results_fileC.txt
              |--- results_fileD.txt
         |--- tool_2/
              |--- results_fileC.txt
              |--- results_fileD.txt
    |--- fastq/
         |--- tool_1/
              |--- results_fileE.txt
              |--- results_fileF.txt
         |--- tool_2/
              |--- results_fileE.txt
              |--- results_fileF.txt
    |--- md5/
         |--- vcf/
              |--- md5_fileA.txt
              |--- md5_fileB.txt
         |--- bam/
              |--- md5_fileC.txt
              |--- md5_fileD.txt
         |--- fastq/
              |--- md5_fileE.txt
              |--- md5_fileF.txt
         |--- other/
              |--- md5_fileG.txt
              |--- md5_fileH.txt
```

Cromwell on AWS
---------------

### Requirements

* S3 bucket for input and output
* EC2 instance for Cromwell server
* EC2 launch template
    - Genomics workloads
    - Interfacing with Cromwell
* AWS Batch
    - Compute environment
    - Job queue

#### S3 bucket

* Created using CFN template 
  [here](https://github.com/aws-samples/aws-genomics-workflows/blob/master/src/templates/aws-genomics-s3.template.yaml)

#### EC2 Launch Template

* Generated using CFN template at 
  [here](https://github.com/aws-samples/aws-genomics-workflows/blob/master/src/templates/aws-genomics-launch-template.template.yaml)

* Has a LaunchTemplateID in the CFN Stack Detail, under Outputs


#### AWS Batch

* Helps you run batch jobs on AWS Cloud (d'oh)
* Submit jobs to the queue, specifying app to run and CPU/mem reqs
* Responsible for launching the appropriate quantity 
  and types of EC2 instances needed to run the jobs

* Manages:
  - Job definitions: Docker, vCPUs, mem etc.
  - Job queues
  - Compute environments: EC2 management
    
* Job queues `<->` one or more compute environments
* Compute environment `<->` one or more job queues


